File: Assets/Scripts/App/Game/Object/Role/Monster/Monster.cs
using System.Collections.Generic;
using App.Data.DataClass.Game.Object;
using App.Game.Generic.BaseObject;
using App.Game.Generic.Map;
using App.UI.GameScene.Control;
using DG.Tweening;
using UnityEngine;
using UnityEngine.EventSystems;
using UnityEngine.UI;

namespace App.Game.Object.Monster
{
    public class Monster : BaseRole, IMonster
    {
        public MonsterData data;

        public float hp;
        public float speed;
        public Animator animator;
        public Transform hpImageBg; // 血条背景图片
        public Transform hpImageFg; // 血条前景图片
        public Transform signFather; // 集火标记父对象

        private readonly List<BaseBuffEffect> buffEffects = new List<BaseBuffEffect>();
        private GameFacade facade;
        private GameManager gameManager;
        private float growth = 1.0f; // 成长系数
        protected float lastWoundTime; // 上次扣血时间

        private Cell nextCell;
        private int pathIndex;

        private List<Cell> pathList;

        protected virtual void Awake()
        {
            animator = GetComponent<Animator>();

            gameManager = GameManager.Instance;
            facade = GameFacade.Instance;
            hpImageBg = transform.Find("HpHolder");
            hpImageFg = transform.Find("HpHolder/HpSlider");
        }

        protected virtual void Start()
        {
            // 默认隐藏血条
            hpImageBg.gameObject.SetActive(false);
        }

        protected virtual void Update()
        {
            Move();

            // 判断是否到达目标格子
            if (Vector3.Distance(Map.Map.GetCellCenterPos(nextCell), transform.position) < 0.1f && IsDead == false)
            {
                // 到达终点格子, 触发死亡方法
                if (pathIndex == pathList.Count - 1)
                {
                    // 终点怪物死亡
                    EndDead();
                    return;
                }

                // 到达换下个目标格子
                pathIndex++;
                pathIndex = Mathf.Clamp(pathIndex, 0, pathList.Count - 1);
                nextCell = pathList[pathIndex];
            }

            // 超过2秒没受到伤害或怪物死亡隐藏血条
            if (Time.time - lastWoundTime > 2 || IsDead) hpImageBg.gameObject.SetActive(false);
        }

        /// <summary>
        ///     点击触发集火
        /// </summary>
        private void OnMouseDown()
        {
            if (IsDead) return;

            // 射线检测判断是否被UI遮挡
            var gr = gameManager.uiManager.canvas.GetComponent<GraphicRaycaster>();
            var eventData = new PointerEventData(EventSystem.current) { position = Input.mousePosition };
            var results = new List<RaycastResult>();
            gr.Raycast(eventData, results);
            // 被显示范围的Ui遮挡除外
            if (results.Count > 0 && results[0].gameObject.name != "ImageAttackRange") return;
            // 将自己的位置信息传出
            gameManager.sceneManager.SetFireTarget(this);
        }

        public virtual void Init(List<Cell> list, float hard, MonsterData data)
        {
            this.data = data;
            pathList = list;
            Growth = hard;
            // 位置设置在起点
            transform.position = Map.Map.GetCellCenterPos(pathList[0]);
            // 设置第一个目标格子
            nextCell = pathList[0];
        }

        public void AddBuffEffect(BaseBuffEffect buffEffect)
        {
            buffEffects.Add(buffEffect);
        }

        public override void Wound(int woundHp)
        {
            Hp -= woundHp;
            if (Hp <= 0)
                // 播放死亡音效
                // 播放死亡音效
                gameManager.PlaySound("Music/MonsterDead", 1, false);
        }

        public virtual void SetSpeed(float v)
        {
            speed = v;
        }

        public virtual Transform GetSignFather()
        {
            return signFather;
        }

        public override void Dead()
        {
            // 回收
            gameManager.poolManager.PushObject(gameObject);
            // 怪物死亡触发判断胜利
            gameManager.eventCenter.TriggerEvent("JudgeWin");
            // 记录到统计信息
            gameManager.sceneManager.UpdateKillMonsterCount(+1);
        }

        public override void OnPush()
        {
            // 清空数据
            nextCell = null;
            // 还原动画
            animator.SetBool("Dead", false);
        }

        /// <summary>
        ///     每次从缓存池取出初始化数据
        /// </summary>
        public override void OnGet()
        {
            pathIndex = 0;
            // 刷新属性
            hp = data.maxHp;
            speed = data.speed;
            IsDead = false;
        }

        private void ClearAllBuffs()
        {
            // 移除身上所有Buff
            for (var i = 0; i < buffEffects.Count; i++) gameManager.poolManager.PushObject(buffEffects[i].gameObject);

            gameManager.buffManager.RemoveAllBuffs(this);
        }

        private void Move()
        {
            if (gameManager.sceneManager.IsPause() || IsDead) return;

            var dir = Map.Map.GetCellCenterPos(nextCell) - transform.position;
            dir.Normalize();
            // 移动
            transform.Translate(dir * (Time.deltaTime * speed));
        }

        /// <summary>
        ///     终点死亡
        /// </summary>
        private void EndDead()
        {
            hp = 0;
            IsDead = true;
            // 触发怪物到达终点事件
            gameManager.sceneManager.Spawner.Carrot.Wound((int)data.atk);
            // 移除所有Buff
            ClearAllBuffs();
            // 播放死亡动画
            animator.SetBool("Dead", true);
            // 播放死亡音效
            gameManager.PlaySound("Music/MonsterDead", 1, false);
        }

        #region 属性

        public float Hp
        {
            get => hp;
            set
            {
                hp = value;
                if (hp <= 0)
                {
                    hp = 0;
                    IsDead = true;
                    // 加钱
                    gameManager.sceneManager.UpdateMoney(+(int)(data.baseMoney * growth));
                    // 生成加钱UI
                    var addMoneyTips =
                        gameManager.factoryManager.UIControlFactory.CreateControl("AddMoneyTips").GetComponent<AddMoneyTips>();
                    addMoneyTips.textMeshPro.text = "+" + (int)(data.baseMoney * growth);
                    addMoneyTips.transform.position = transform.position;
                    addMoneyTips.transform.DOMoveY(addMoneyTips.transform.position.y + 2f, 0.5f); // 上移动画
                    // 移除所有Buff
                    ClearAllBuffs();
                    // 如果集火的是自己取消集火标志
                    if ((Monster)gameManager.sceneManager.Spawner.GetCollectingFiresTarget() == this) gameManager.sceneManager.CancelFire();

                    // 播放死亡动画
                    animator.SetBool("Dead", true);
                }

                // 更新血条图片
                hpImageBg.gameObject.SetActive(true);
                hpImageFg.localScale = new Vector3(hp / (data.maxHp * (growth == 0 ? 1 : growth)), 1, 1);
                // 记录显示血条的时间
                lastWoundTime = Time.time;
            }
        }

        public float Growth
        {
            get => growth;
            set
            {
                growth = value;
                // 修改成长系数时自动修改属性值
                hp *= growth;
            }
        }

        public MonsterData Data => data;

        #endregion
    }
}

File: Assets/Scripts/App/Lua/App/Object/Monster.lua
Monster = Object:SubClass('Monster')

Monster.cs = {}
Monster.mono = {}
Monster.obj = {}
Monster.gameManager = nil

Monster.data = null
Monster.hp = 0
Monster.isDead = false
Monster.growth = 1
Monster.speed = 0

Monster.nextCell = nil
Monster.pathIndex = 0
Monster.pathList = nil
-- 血条
Monster.hpBg = nil
Monster.hpFg = nil
Monster.lastWoundTime = 0
-- 集火标志
Monster.signFather = nil
Monster.animator = nil
Monster.canMove = false
Monster.buffsList = nil

function Monster:Construct(monsterObj)
    self.gameManager = GameManager.Instance
    self.buffsList = List:New()
    self.obj = monsterObj

    if self.obj:GetComponent(typeof(LuaMonster)) == nil then
        self.cs = self.obj:AddComponent(typeof(LuaMonster))
    else
        self.cs = self.obj:GetComponent(typeof(LuaMonster))
    end

    if self.obj:GetComponent(typeof(MonoScript)) == nil then
        self.mono = self.obj:AddComponent(typeof(MonoScript))
    else
        self.mono = self.obj:GetComponent(typeof(MonoScript))
    end

    self:InitCsAction()

    self.animator = self.obj:GetComponent('Animator')
    self.hpBg = self.obj.transform:Find('HpHolder')
    self.hpFg = self.obj.transform:Find('HpHolder/HpSlider')
    self.signFather = self.obj.transform:Find('SignFather')

    -- 默认隐藏血条
    self.hpBg.gameObject:SetActive(false)
end

function Monster:InitCsAction()
    self.cs.onGetHpAction = function()
        return self:GetHp()
    end
    self.cs.onSetHpAction = function(v)
        self:SetHP(v)
    end
    self.cs.onGetGrowthAction = function()
        return self:GetGrowth()
    end
    self.cs.onSetGrowthAction = function(v)
        return self:SetGrowth(v)
    end
    self.cs.onGetIsDeadAction = function()
        return self:GetIsDead()
    end
    self.cs.onSetIsDeadAction = function(isDead)
        self:SetIsDead(isDead)
    end
    self.cs.onGetDataAction = function()
        return self:GetData()
    end
    self.cs.onGetTransformAction = function()
        return self:GetTransfrom()
    end
    self.cs.onWoundAction = function(v)
        self:Wound(v)
    end
    self.cs.onPushAction = function()
        self:OnPush()
    end
    self.cs.onGetAction = function()
        self:OnGet()
    end
    self.cs.onInitAction = function(cellList, hard, data)
        self:Init(cellList, hard, data)
    end
    self.cs.onSetSpeedAction = function(v)
        self:SetSpeed(v)
    end
    self.cs.onGetSignFatherAction = function()
        return self:GetSignFather()
    end
    self.cs.onDeadAction = function()
        self:Dead()
    end
    self.cs.onAddBuffEffectAction = function(buffEffct)
        self:AddBuffEffect(buffEffct)
    end
    self.cs.onMouseDownAction = function()
        self:OnMouseDown()
    end

    -- unity 回调
    self.mono.onUpdateAction = function()
        self:Update()
    end
end

-- unity回调
function Monster:Update()
    if self.canMove == true then
        self:Move()
    end
    -- 自动隐藏血条
    if Time.time - self.lastWoundTime > 2 or self.isDead == true then
        self.hpBg.gameObject:SetActive(false)
    end
end

function Monster:OnMouseDown()
    if self.isDead == true then
        return
    end

    local gr = self.gameManager.uiManager.canvas:GetComponent('GraphicRaycaster')
    local eventData = CS.UnityEngine.EventSystems.PointerEventData(CS.UnityEngine.EventSystems.EventSystem.current)
    eventData.position = Input.mousePosition
    local results = CS.System.Collections.Generic.List(CS.UnityEngine.EventSystems.RaycastResult)()
    gr:Raycast(eventData, results)
    if results.Count > 0 and results[0].gameObject.name ~= 'ImageAttackRange' then
        return
    end

    self.gameManager.sceneManager:SetFireTarget(self.cs)
end

-- 移动
function Monster:Move()
    -- 暂停或死亡禁止移动
    if self.gameManager.sceneManager:IsPause() == true or self.isDead == true then
        return
    end
    if self.nextCell == nil then
        print(self.isDead)
    end

    -- 判断是否到达格子
    local distance = Vector3.Distance(CSMap.GetCellCenterPos(self.nextCell), self.obj.transform.position)
    if distance < 0.1 and self.isDead == false then
        -- 下一个格子
        self.pathIndex = self.pathIndex + 1
        if self.pathIndex > self.pathList.Count - 1 then
            -- 到达终点重新循环
            self.pathIndex = 0
        end
        self.nextCell = self.pathList[self.pathIndex]
    end

    local dir = CSMap.GetCellCenterPos(self.nextCell) - self.obj.transform.position
    dir:Normalize()
    self.obj.transform:Translate(dir * Time.deltaTime * self.speed)
end

-- 重写
function Monster:Init(cellList, hard, data)
    self.data = data
    self.growth = hard
    self.pathList = cellList

    self.hp = self.data.maxHp * self.growth
    self.speed = self.data.speed
    -- 设置在开头
    self.obj.transform.position = CSMap.GetCellCenterPos(self.pathList[0])
    self.pathIndex = 0
    self.nextCell = self.pathList[0]
    self.canMove = true
    self.isDead = false
end

function Monster:GetHp()
    return self.hp
end

function Monster:SetHP(v)
    self.hp = v
end

function Monster:GetGrowth()
    return self.growth
end

function Monster:SetGrowth(v)
    self.growth = v
end

function Monster:GetIsDead()
    return self.isDead
end

function Monster:SetIsDead(isDead)
    self.isDead = isDead
end

function Monster:GetData()
    return self.data
end

function Monster:GetTransfrom()
    return self.obj.transform
end

function Monster:Wound(v)
    self.hp = self.hp - v
    -- 更新血条
    self.hpBg.gameObject:SetActive(true)
    local maxHp = self.data.maxHp * self.growth
    self.hpFg.localScale = Vector3(self.hp / maxHp, 1, 1)
    self.lastWoundTime = Time.time

    if self.hp <= 0 then
        self.hp = 0
        self.isDead = true
        self.gameManager.sceneManager:UpdateMoney(math.floor(self.data.baseMoney * self.growth))
        -- 加钱ui
        local ui =
            GameManager.Instance.factoryManager.UIControlFactory:CreateControl('AddMoneyTips'):GetComponent(
            'AddMoneyTips'
        )
        ui.textMeshPro.text = '+' .. math.floor(self.data.baseMoney * self.growth)
        ui.transform.position = self.mono.transform.position
        CS.DG.Tweening.ShortcutExtensions.DOMoveY(ui.transform, ui.transform.position.y + 2, 0.5)
        -- 移除buff
        self:ClearAllBuffs()

        if self.gameManager.sceneManager.Spawner:GetCollectingFiresTarget() == self.cs then
            self.gameManager.sceneManager:CancelFire()
        end
        -- 触发死亡动画
        self.animator:SetBool('Dead', true)
    end
end

function Monster:ClearAllBuffs()
    for i = 0, self.buffsList.count - 1 do
        self.gameManager.poolManager:PushObject(self.buffsList:Get(i).gameObject)
    end
    self.gameManager.buffManager:RemoveAllBuffs(self.cs)
    self.buffsList:Clear()
end

function Monster:AddBuffEffect(buffEffct)
    self.buffsList:Add(buffEffct)
end

function Monster:Dead()
    self.gameManager.poolManager:PushObject(self.mono.gameObject)
    self.gameManager.eventCenter:TriggerEvent('JudgeWin')
    self.gameManager.sceneManager:UpdateKillMonsterCount(1)
end

function Monster:OnPush()
    self.nextCell = nil
    self.animator:SetBool('Dead', false)
end

function Monster:OnGet()
    self.pathIndex = 0
end

function Monster:SetSpeed(v)
    self.speed = v
end

function Monster:GetSignFather()
    return self.signFather
end
